server/src/main/java/org/apache/iotdb/db/engine/compaction/performer/impl/ReadPointCompactionPerformer.java

```diff
@@ -239,6 +238,60 @@ private void compactNonAlignedSeries(
     compactionWriter.endChunkGroup();
   }
 
+  private ConcurrentHashMap<String, MeasurementSchema> getMeasurementSchema(
+      String device, Set<String> measurements) throws IllegalPathException, IOException {
+    ConcurrentHashMap<String, MeasurementSchema> schemaMap = new ConcurrentHashMap<>();
+    List<TsFileResource> allResources = new LinkedList<>(seqFiles);
+    allResources.addAll(unseqFiles);
+    for (String measurement : measurements) {
+      for (TsFileResource tsFileResource : allResources) {
+        if (!tsFileResource.mayContainsDevice(device)) {
+          continue;
+        }
+        MeasurementSchema schema =
+            getMeasurementSchemaFromReader(
+                readerCacheMap.computeIfAbsent(
+                    tsFileResource,
+                    x -> {
+                      try {
+                        return new TsFileSequenceReader(x.getTsFilePath());
+                      } catch (IOException e) {
+                        throw new RuntimeException(
+                            String.format(
+                                "Failed to construct sequence reader for %s", tsFileResource));
+                      }
+                    }),
+                device,
+                measurement);
+        if (schema != null) {
+          schemaMap.put(measurement, schema);
+          break;
+        }
+      }
+    }
+    return schemaMap;
+  }
+
+  private MeasurementSchema getMeasurementSchemaFromReader(
+      TsFileSequenceReader reader, String device, String measurement)
+      throws IllegalPathException, IOException {
+    List<ChunkMetadata> chunkMetadata =
+        reader.getChunkMetadataList(new PartialPath(device, measurement));
+    if (chunkMetadata.size() > 0) {
+      Chunk chunk = reader.readMemChunk(chunkMetadata.get(0));
+      ChunkHeader header = chunk.getHeader();
+      return new MeasurementSchema(
+          measurement, header.getDataType(), header.getEncodingType(), header.getCompressionType());
+    }
+    return null;
+  }
```

Comment:
Reading all ChunkMetadata and the one chunk just for the MeasurementSchema seems too costly, and I think the design should be revised.